language-support/hs/bindings/src/DA/Ledger/Services/TransactionService.hs

```diff
@@ -30,112 +31,89 @@ getTransactions req =
             let LL.TransactionService {transactionServiceGetTransactions=rpc} = service
-            sendToStreamFlat timeout (lowerRequest req) f stream rpc
+            GRPC.sendToStreamFlat timeout (lowerRequest req) f stream rpc
     return stream
-    where f = raiseList raiseTransaction . LL.getTransactionsResponseTransactions
+    where f = Convert.raiseList Convert.raiseTransaction . LL.getTransactionsResponseTransactions
 
 getTransactionTrees :: GetTransactionsRequest -> LedgerService (Stream TransactionTree)
 getTransactionTrees req =
     makeLedgerService $ \timeout config -> do
     stream <- newStream
     _ <- forkIO $
-        withGRPCClient config $ \client -> do
+        HL.withGRPCClient config $ \client -> do
             service <- LL.transactionServiceClient client
             let LL.TransactionService {transactionServiceGetTransactionTrees=rpc} = service
-            sendToStreamFlat timeout (lowerRequest req) f stream rpc
+            GRPC.sendToStreamFlat timeout (lowerRequest req) f stream rpc
     return stream
-    where f = raiseList raiseTransactionTree . LL.getTransactionTreesResponseTransactions
+    where f = Convert.raiseList Convert.raiseTransactionTree . LL.getTransactionTreesResponseTransactions
+
+getTransactionBy :: (LL.TransactionService HL.ClientRequest HL.ClientResult
+                        -> HL.ClientRequest 'HL.Normal request response
+                        -> IO (HL.ClientResult stream response))
+                    -> (LedgerId -> tx_ref -> [Party] -> request)
+                    -> (response -> Convert.Perhaps tx)
+                    -> (LedgerId -> tx_ref -> [Party] -> LedgerService (Maybe tx))
+getTransactionBy get_rpc mk_request resp_handler
+                 ledger_id transaction_ref parties
+    = makeLedgerService $ \timeout config -> do
+      HL.withGRPCClient config $ \client -> do
+          service <- LL.transactionServiceClient client
+          let rpc = get_rpc service
+          rpc (HL.ClientNormalRequest (mk_request ledger_id transaction_ref parties) timeout GRPC.emptyMdm)
+          >>= \case
+              HL.ClientNormalResponse resp _m1 _m2 _status _details -> case resp_handler resp of
+                                                                         Left reason -> fail (show reason)
+                                                                         Right x -> return $ Just x
+              HL.ClientErrorResponse (HL.ClientIOError (HL.GRPCIOBadStatusCode HL.StatusNotFound _details)) ->
+                  return Nothing
+              HL.ClientErrorResponse e ->
+                  fail (show e)
+              HL.ClientWriterResponse _ _ _ _ _ -> fail ""
+              HL.ClientReaderResponse _ _ _ -> fail ""
+              HL.ClientBiDiResponse _ _ _ -> fail ""
+
+extractTxR :: LL.GetTransactionResponse -> Convert.Perhaps TransactionTree
+extractTxR (LL.GetTransactionResponse Nothing) = fail "GetTransactionResponse, transaction field is missing"
```

Comment:
This does not match the previous behavior. Previously you called `fail` in `IO` which ends up throwing an `IOException`. Now you are calling it in `Either e a` which ends up calling `error`. Given that exceptions thrown in `IO` are usually a lot easier to reason about (you don’t have to worry about laziness and when it will actually be thrown), I would suggest one of the following:
1. Either go via the `e` part of `Either e a`, i.e., return something like `Left (Missing "GetTransactionResponse, transaction field is missing")`.
2. Throw exceptions via `throwIO`. This will require generalizing `getTransactionBy` to allow these conversions to happen in `IO`.